src/librustc/infer/error_reporting/mod.rs

```diff
@@ -1155,6 +1158,87 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             _ => {}
         }
 
+        struct OpaqueTypesVisitor<'tcx> {
+            types: FxHashMap<&'static str, FxHashSet<Span>>,
+            expected: FxHashMap<&'static str, FxHashSet<Span>>,
+            found: FxHashMap<&'static str, FxHashSet<Span>>,
+            ignore_span: Span,
+            tcx: TyCtxt<'tcx>,
+        }
+
+        impl<'tcx> OpaqueTypesVisitor<'tcx> {
+            fn visit_expected_found(
+                tcx: TyCtxt<'tcx>,
+                expected: Ty<'tcx>,
+                found: Ty<'tcx>,
+                ignore_span: Span,
+            ) -> Self {
+                let mut types_visitor = OpaqueTypesVisitor {
+                    types: Default::default(),
+                    expected: Default::default(),
+                    found: Default::default(),
+                    ignore_span,
+                    tcx,
+                };
+                expected.visit_with(&mut types_visitor);
+                std::mem::swap(&mut types_visitor.expected, &mut types_visitor.types);
+                found.visit_with(&mut types_visitor);
+                std::mem::swap(&mut types_visitor.found, &mut types_visitor.types);
+                types_visitor
+            }
+
+            fn report(&self, err: &mut DiagnosticBuilder<'_>) {
+                for (target, types) in &[("expected", &self.expected), ("found", &self.found)] {
+                    for (key, values) in types.iter() {
+                        let count = values.len();
+                        for sp in values {
+                            err.span_label(*sp, format!(
+                                "{}this is {}the {} {}{}",
+                                if sp.is_desugaring(DesugaringKind::Async) {
+                                    "in the desugared `async fn`, "
+                                } else {
+                                    ""
+                                },
+                                if count > 1 {
+                                    "one of"
+                                } else {
+                                    ""
+                                },
+                                target,
+                                key,
+                                pluralize!(count),
+                            ));
+                        }
+                    }
+                }
+            }
+        }
+
+        impl<'tcx> ty::fold::TypeVisitor<'tcx> for OpaqueTypesVisitor<'tcx> {
+            fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+                let kind = match t.kind {
+                    ty::Closure(..) => "closure",
```

Comment:
```suggestion
                    ty::Closure(def_id, _) => (def_id, &mut self.types_closure, &mut self.expected_closure),
```

Reply:
I'm not sure what the suggestion was for here.

Reply:
This is related to the discussion in the other comment above.